### PR TITLE
Move to Error step if ClusterState* steps throw

### PR DESCRIPTION
**Pull request by AthenaEryma**
Previously, if ClusterStateActionSteps or ClusterStateWaitSteps threw an
exception while executing, the exception would only be caught and logged by
the generic ClusterStateUpdateTask machinery and the index would become
stuck on that step.

Now, exceptions thrown in these steps will be caught and the index will
be moved to the Error step.

Credit to @talevy for the test cases, turns out we both worked on this separately and he got to them first.

### CODE REVIEW

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/ExecuteStepsUpdateTask.java [hunk 1]
```diff
@@ -172,4 +181,11 @@ public void onFailure(String source, Exception e) {
         throw new ElasticsearchException(
                 "policy [" + policy + "] for index [" + index.getName() + "] failed on step [" + startStep.getKey() + "].", e);
     }
+
+    private ClusterState moveToErrorStep(final ClusterState state, Step.StepKey currentStepKey, Exception cause) throws IOException {
```
Comment (talevy) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/ExecuteStepsUpdateTask.java]: this should probably be RuntimeException here?

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/indexlifecycle/ExecuteStepsUpdateTaskTests.java [hunk 2]
```diff
@@ -253,6 +254,43 @@ public void testOnFailure() throws IOException {
...
+        LifecycleExecutionState lifecycleState = LifecycleExecutionState.fromIndexMetadata(newState.getMetaData().index(index));
+        StepKey currentStepKey = IndexLifecycleRunner.getCurrentStepKey(lifecycleState);
+        assertThat(currentStepKey, equalTo(new StepKey(firstStepKey.getPhase(), firstStepKey.getAction(), ErrorStep.NAME)));
+        assertThat(firstStep.getExecuteCount(), equalTo(1L));
+        assertThat(secondStep.getExecuteCount(), equalTo(0L));
+        assertThat(task.getNextStepKey(), equalTo(secondStep.getKey()));
+        assertThat(lifecycleState.getPhaseTime(), nullValue());
+        assertThat(lifecycleState.getActionTime(), nullValue());
+        assertThat(lifecycleState.getStepInfo(), equalTo("{\"type\":\"runtime_exception\",\"reason\":\"error\"}"));
+    }
```
Comment (talevy) [inline x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/indexlifecycle/ExecuteStepsUpdateTaskTests.java]: nit: newline

**Review by talevy — APPROVED** [review]
left some comments I think should be addressed

**Review by dakrone — CHANGES_REQUESTED** [review]
I left a question

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/ExecuteStepsUpdateTask.java [hunk 3]
```diff
@@ -92,7 +92,11 @@ public ClusterState execute(final ClusterState currentState) throws IOException
                     // move the cluster state to the next step
                     logger.trace("[{}] performing cluster state action ({}) [{}], next: [{}]",
                         index.getName(), currentStep.getClass().getSimpleName(), currentStep.getKey(), currentStep.getNextStepKey());
-                    state = ((ClusterStateActionStep) currentStep).performAction(index, state);
+                    try {
+                        state = ((ClusterStateActionStep) currentStep).performAction(index, state);
+                    } catch (RuntimeException exception) {
```
[the remark this answers is not in the document]
Comment (gwbrown) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/ExecuteStepsUpdateTask.java]: This was left over from an earlier draft of this that had more in the `try` block, and I wasn't sure if we wanted to include `IOException` in this catch. I've changed it.

**Review by dakrone — APPROVED** [review]

**Pull request merged**